py/gc.c

```diff
@@ -113,8 +113,8 @@
 #endif
 
 #if MICROPY_PY_THREAD && !MICROPY_PY_THREAD_GIL
-#define GC_ENTER() mp_thread_mutex_lock(&MP_STATE_MEM(gc_mutex), 1)
-#define GC_EXIT() mp_thread_mutex_unlock(&MP_STATE_MEM(gc_mutex))
+#define GC_ENTER() mp_thread_recursive_mutex_lock(&MP_STATE_MEM(gc_mutex), 1)
```

Comment:
I suggest adding a `GC_MUTEX_INIT` macro here, and using it below in `gc_init`, so that the condition `MICROPY_PY_THREAD && !MICROPY_PY_THREAD_GIL` doesn't need to be repeated below (otherwise the init and use of the mutex may get out of sync wrt which config options it's enabled for).

Reply:
Good idea, have made the change